### PR TITLE
Raise NotImplementedError on missing backend functionality

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #2053 

~~Note I've been having mkl errors when using `.summary()`. It would be nice to find a non-mkl workaround `if not torch.backends.mkl.is_available()`.~~

## Tested
- added a unit test

### CODE REVIEW

**Comment by neerajprad** [comment]
> if not torch.backends.mkl.is_available()

I think the diagnostics should just have `nan` if mkl isn't available (that's probably required by the fft function), and IIRC we made a change to do that sometime back. This looks like a regression to me. cc. @fehiepsi

File: tests/test_generic.py [hunk 1]
```diff
@@ -16,4 +16,14 @@ def test_mcmc_interface(model, backend):
...
+            mcmc.summary()
+
+
+@pytest.mark.parametrize('backend', ['pyro', 'minipyro'])
+def test_not_implemented(backend):
+    with pyro_backend(backend):
+        pyro.sample  # should be implemented
+        pyro.param  # should be implemented
+        with pytest.raises(NotImplementedError):
+            pyro.nonexistant_primitive
```
Comment (neerajprad) [inline tests/test_generic.py]: nit: `existent` misspelt.

**Comment by neerajprad** [comment]
LGTM! I'll fix the mkl issue in another PR.

**Comment by neerajprad** [comment]
@fritzo - could you paste the error trace? I think the only diagnostic function that is likely to throw this is `effective_sample_size` but that is wrapped in a _safe block. I think the issue is that perhaps pytest is treating that warning as an error? If that's the case, we could just add `@pytest.mark.filterwarnings("ignore:..")` to the test. 

```
def _safe(fn):
    """
    Safe version of utilities in the :mod:`pyro.ops.stats` module. Wrapped
    functions return `NaN` tensors instead of throwing exceptions.

    :param fn: stats function from :mod:`pyro.ops.stats` module.
    """
    @functools.wraps(fn)
    def wrapped(sample, *args, **kwargs):
        try:
            val = fn(sample, *args, **kwargs)
        except Exception:
            warnings.warn(tb.format_exc())
            val = torch.full(sample.shape[2:], float("nan"),
                             dtype=sample.dtype, device=sample.device)
        return val

    return wrapped
```

**Comment by fritzo** [comment]
Thanks, @neerajprad your suggestion of `@pytest.mark.filterwarnings` fixed my issue.

**Pull request merged**